drivers/wireless/cyw43/bus.zig

```diff
@@ -0,0 +1,370 @@
+
+        self.backplane_set_window(addr);
+
+        var bus_addr = addr & consts.BACKPLANE_ADDRESS_MASK;
+        if (len == 4) {
+            bus_addr |= consts.BACKPLANE_ADDRESS_32BIT_FLAG;
+        }
+
+        self.writen(.backplane, @truncate(bus_addr), value, len);
+    }
+
+    fn backplane_set_window(self: *Self, addr: u32) void {
+        const new_window = addr & ~consts.BACKPLANE_ADDRESS_MASK;
+
+        if (@as(u8, @truncate(new_window >> 24)) != @as(u8, @truncate(self.backplane_window >> 24))) {
+            self.write8(.backplane, consts.REG_BACKPLANE_BACKPLANE_ADDRESS_HIGH, @as(u8, @truncate(new_window >> 24)));
+        }
+        if (@as(u8, @truncate(new_window >> 16)) != @as(u8, @truncate(self.backplane_window >> 16))) {
+            self.write8(.backplane, consts.REG_BACKPLANE_BACKPLANE_ADDRESS_MID, @as(u8, @truncate(new_window >> 16)));
+        }
+        if (@as(u8, @truncate(new_window >> 8)) != @as(u8, @truncate(self.backplane_window >> 8))) {
+            self.write8(.backplane, consts.REG_BACKPLANE_BACKPLANE_ADDRESS_LOW, @as(u8, @truncate(new_window >> 8)));
+        }
+
+        self.backplane_window = new_window;
+    }
+
+    fn read32_swapped(self: *Self, func: FuncType, addr: u17) u32 {
+        const cmd = Cyw43Cmd{ .cmd = .read, .incr = .incremental, .func = func, .addr = addr, .len = 4 };
+        const cmd_swapped = swap16(@bitCast(cmd));
+
+        var buff = [1]u32{0};
+        _ = self.spi.spi_read_blocking(cmd_swapped, &buff);
+
+        return swap16(buff[0]);
+    }
+
+    fn write32_swapped(self: *Self, func: FuncType, addr: u17, value: u32) void {
+        const cmd = Cyw43Cmd{ .cmd = .write, .incr = .incremental, .func = func, .addr = addr, .len = 4 };
+
+        var buff: [2]u32 = .{ swap16(@bitCast(cmd)), swap16(value) };
+        _ = self.spi.spi_write_blocking(&buff);
+    }
+
+    fn swap16(x: u32) u32 {
+        return x << 16 | x >> 16;
+    }
+};
+
+const CmdType = enum(u1) {
+    read = 0,
+    write = 1,
+};
+
+const IncrMode = enum(u1) {
+    fixed = 0, // Fixed address (no increment)
+    incremental = 1, // Incremental burst
+};
+
+const FuncType = enum(u2) { bus = 0, backplane = 1, wlan = 2, bt = 3 };
+
+const Cyw43Cmd = packed struct(u32) {
+    len: u11,
+    addr: u17,
+    func: FuncType = .bus,
+    incr: IncrMode = .fixed,
+    cmd: CmdType,
+};
+
+const CtrlWordLength = enum(u1) { word_16 = 0, word_32 = 1 };
+
+const CtrlEndianess = enum(u1) { little_endian = 0, big_endian = 1 };
+
+const CtrlSpeedMode = enum(u1) { normal = 0, high_speed = 1 };
+
+const CtrlInterruptPolarity = enum(u1) { low_polarity = 0, high_polarity = 1 };
+
+const CtrlReg = packed struct(u8) {
+    word_length: CtrlWordLength,
+    endianess: CtrlEndianess,
+    reserved1: u2 = 0,
+    speed_mode: CtrlSpeedMode,
+    interrupt_polarity: CtrlInterruptPolarity,
+    reserved3: u1 = 0,
+    wake_up: bool,
+};
+
+const ResponseDelay = packed struct(u8) {
```

Comment:
should these be extern structs?

Reply:
Could be, but in fact this is bit field but with some unknown flags so to keep that information I choosed packed struct